src/block_neural.rs

```diff
@@ -358,94 +358,92 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockNeuronLayer<L> {
+                        let gradient = general_gradient * feature_value;
+                        let update = self.optimizer.calculate_update(
+                            gradient,
+                            &mut self
+                                .weights_optimizer
+                                .get_unchecked_mut(i + j_offset)
+                                .optimizer_data,
+                        );
+                        *output_errors.get_unchecked_mut(i) +=
+                            self.weights.get_unchecked(i + j_offset).weight * general_gradient;
+                        self.weights.get_unchecked_mut(i + j_offset).weight -= update;
+                    }
+                    {
+                        // Updating bias term:
+                        let gradient = general_gradient * 1.0;
+                        let update = self.optimizer.calculate_update(
+                            gradient,
+                            &mut self
+                                .weights_optimizer
+                                .get_unchecked_mut(bias_offset + j)
+                                .optimizer_data,
+                        );
+                        self.weights.get_unchecked_mut(bias_offset + j).weight -= update;
+                    }
+
+                    if self.max_norm != 0.0 && fb.example_number % 10 == 0 {
+                        let mut wsquaredsum = 0.000001; // Epsilon
+                        for i in 0..self.num_inputs {
+                            let w = self.weights.get_unchecked_mut(i + j_offset).weight;
+                            wsquaredsum += w * w;
                         }
-                        let var1 = (sumsqr - sum * sum / bias_offset as f32) / bias_offset as f32;
-                        let var2 = var1.sqrt();
-                        for i in 0..bias_offset {
-                            self.weights.get_unchecked_mut(i).weight /= var2;
+                        let norm = wsquaredsum.sqrt();
+                        if norm > self.max_norm {
+                            let scaling = self.max_norm / norm;
+                            for i in 0..self.num_inputs {
+                                self.weights.get_unchecked_mut(i + j_offset).weight *= scaling;
+                            }
                         }
                     }
-
-                    input_tape.copy_from_slice(output_errors.get_unchecked(0..self.num_inputs));
                 }
+                if self.layer_norm && fb.example_number % 10 == 0 {
+                    let mut sum: f32 = 0.0;
+                    let mut sumsqr: f32 = 0.0;
+                    let K = 100.0;
+                    for i in 0..bias_offset {
+                        let w = self.weights.get_unchecked(i).weight - K;
+                        sum += w;
+                        sumsqr += w * w;
+                    }
+                    let var1 = (sumsqr - sum * sum / bias_offset as f32) / bias_offset as f32;
```

Comment:
Please also rename var1 & var2 to something more appropriate

Reply:
Thanks, we actually need only one if v1 is sqrt-ed already (it's the only use anyways) -> 97f259f6841325315c91e090c34e3da0f7c80fda